File: aggerators/agg.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <limits>
#include <chrono>
#include <vector>

struct CityData {
    double minTemp = std::numeric_limits<double>::max();
    double maxTemp = std::numeric_limits<double>::lowest();
    double totalTemp = 0;
    int count = 0;

    void update(double temperature) {
        minTemp = std::min(minTemp, temperature);
        maxTemp = std::max(maxTemp, temperature);
        totalTemp += temperature;
        count++;
    }
    
    std::string toString() const {
        return "Min: " + std::to_string(minTemp) + 
               " / Mean: " + std::to_string(totalTemp / count) + 
               " / Max: " + std::to_string(maxTemp);
    }
};

void aggregate() {
    std::unordered_map<std::string, CityData> data;
    std::ifstream file("../measurements.txt", std::ios::binary);

    std::string line, name;
    double temperature;
    std::vector<char> buffer(100 * 1024 * 1024); // 200MB buffer
    file.rdbuf()->pubsetbuf(buffer.data(), buffer.size());

    while (std::getline(file, line)) {
        if (line.empty()) continue;
        
        auto delimiterPos = line.find(';');
        name = line.substr(0, delimiterPos);
        temperature = std::stod(line.substr(delimiterPos + 1));

        data[name].update(temperature);
    }

    for(const auto& elem : data) {
        std::cout << elem.first << ":" << elem.second.toString() << ", ";
    }
}

int main() {
    auto start = std::chrono::high_resolution_clock::now();
    aggregate();
    auto end = std::chrono::high_resolution_clock::now();

    std::chrono::duration<double> diff = end - start;
    std::cout << "Aggregation took " << diff.count() << " seconds" << std::endl;

    return 0;
}

File: aggerators/agg.py
import time
from collections import defaultdict
import pandas as pd
import array


def aggregate():
    data = defaultdict(lambda: {'min': float('+inf'), 'max': float('-inf'), 'sum': 0, 'count': 0})
    # data = defaultdict(lambda: [array.array('f', [float('+inf'), float('-inf'), 0]), array.array('i', [0])])
    data_get = data.__getitem__
    
    with open('../measurements.txt', 'r', buffering=1000000) as f:
        for line in f:
            if not line.strip():
                continue
            name, temperature = line.split(';')
            temperature = float(temperature)
            city = data_get(name)
            # city = data[name]

            city['min'] = min(city['min'], temperature)
            city['max'] = max(city['max'], temperature)
            city['sum'] += temperature
            city['count'] += 1
            # city[0][0] = min(city[0][0], temperature)
            # city[0][1] = max(city[0][1], temperature)
            # city[0][2] += temperature
            # city[1][0] += 1

    # result = []
    for k in data:
        v = data[k]
        data[k] = f"{k}={v['min']:.2f}/{v['sum'] / v['count']:.2f}/{v['max']:.2f}"
        # data[k] = f"{k}={v[0][0]:.2f}/{v[0][2] / v[1][0]:.2f}/{v[0][1]:.2f}"
    print(data)


def aggregate_with_pandas():
    df = pd.read_csv('../measurements.txt', sep=';', header=None, names=['City', 'Temperature'])
    result = df.groupby('City')['Temperature'].agg(['min', 'max', 'mean'])


def main():
    # start = time.time()
    # aggregate_with_pandas()
    # print(f"Aggregation with pandas took {time.time() - start} seconds")
    
    start = time.time()
    aggregate()
    print(f"Aggregation took {time.time() - start} seconds")


main()